File: SockstoreNet/Application/Query/ProductQuery.cs
using Application.Ports;
using Vocabulary;

namespace Application.Query;

public interface IProductQuery
{
    Task<Product?> FindById(ProductId id, CancellationToken cancellationToken);
    Task<IEnumerable<Product>> FindAll(CancellationToken cancellationToken);
}

public class ProductQuery(IProductPort productPort) : IProductQuery
{
    public async Task<Product?> FindById(ProductId id, CancellationToken cancellationToken)
    {
        var product = await productPort.FindById(id, cancellationToken);
        return product?.ToProduct();
    }

    public async Task<IEnumerable<Product>> FindAll(CancellationToken cancellationToken)
    {
        var products = await productPort.FindAll(cancellationToken);
        return products.Select(aggregate => aggregate.ToProduct());
    }
}

File: SockStoreJava/src/application/api/src/main/java/be/itenium/sockstore/api/ProductQuery.java
package be.itenium.sockstore.api;

import be.itenium.sockstore.vocabulary.Product;
import be.itenium.sockstore.vocabulary.ProductId;

import java.util.List;
import java.util.Optional;

public interface ProductQuery {
    Optional<Product> findById(ProductId id);
    List<Product> findAll();
}